File: profiles/java-spring/scaffold/src/test/java/com/example/app/GreetingIntegrationTest.java
package com.example.app;

import static org.assertj.core.api.Assertions.assertThat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;

/**
 * Integration: the flag seam + telemetry wiring THROUGH the running Spring server.
 *
 * <p>Unlike the unit tests (which exercise a class in isolation), this boots the FULL application
 * context on a random port ({@link SpringBootTest.WebEnvironment#RANDOM_PORT}) and drives it over
 * real HTTP with the JDK {@link HttpClient} (which — unlike {@code TestRestTemplate}'s default
 * factory — supports every method incl. PATCH/HEAD/OPTIONS). It measures {@link
 * AppInstrumentationFilter}, {@link GreetingController}, {@link MetricsController}, and {@link
 * NotFoundAdvice} together. The live-flip case is the ★ load-bearing proof that the provider seam
 * reaches the REAL endpoint with NO restart.
 */
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class GreetingIntegrationTest {

  /** The four security headers expected — exactly once each — on EVERY response. */
  private static final Map<String, String> EXPECTED_SECURITY_HEADERS =
      Map.of(
          "X-Content-Type-Options", "nosniff",
          "X-Frame-Options", "DENY",
          "Content-Security-Policy", "default-src 'none'",
          "Referrer-Policy", "no-referrer");

  private static final ObjectMapper MAPPER = new ObjectMapper();
  private final HttpClient client =
      HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build();

  @LocalServerPort private int port;

  @AfterEach
  void restoreEnvFloor() {
    // A live-flip test installs a file provider into the shared static seam — restore the env floor
    // so it cannot leak into a later test.
    FeatureFlags.resetProvider();
  }

  private String base() {
    return "http://localhost:" + port;
  }

  private HttpResponse<String> send(String method, String path, Map<String, String> headers)
      throws IOException, InterruptedException {
    HttpRequest.Builder builder =
        HttpRequest.newBuilder(URI.create(base() + path))
            .timeout(Duration.ofSeconds(5))
            .method(method, HttpRequest.BodyPublishers.noBody());
    headers.forEach(builder::header);
    return client.send(builder.build(), HttpResponse.BodyHandlers.ofString());
  }

  private HttpResponse<String> get(String path) throws IOException, InterruptedException {
    return send("GET", path, Map.of());
  }

  @Test
  void greetingFlagOffServesDefaultBody() throws Exception {
    FeatureFlags.resetProvider();
    HttpResponse<String> response = get("/greeting");

    assertThat(response.statusCode()).isEqualTo(200);
    assertThat(response.headers().firstValue("Content-Type").orElse(""))
        .contains("application/json");
    assertThat(response.body()).isEqualTo("{\"greeting\":\"Hello, world!\"}");
  }

  @Test
  void greetingFlagOnServesNewBody() throws Exception {
    // Install a provider that enables new_greeting on the SAME running server (own-key-only).
    FeatureFlags.setProvider(name -> "new_greeting".equals(name));
    HttpResponse<String> response = get("/greeting");

    assertThat(response.statusCode()).isEqualTo(200);
    assertThat(response.body()).isEqualTo("{\"greeting\":\"Hello, world! (new)\"}");
  }

  @Test
  void healthzReturnsOkBody() throws Exception {
    HttpResponse<String> response = get("/healthz");

    assertThat(response.statusCode()).isEqualTo(200);
    assertThat(response.body()).isEqualTo("{\"status\":\"ok\"}");
  }

  @Test
  void metricsExposesPrometheusCounter() throws Exception {
    get("/greeting"); // record at least one request first
    HttpResponse<String> response = get("/metrics");

    assertThat(response.statusCode()).isEqualTo(200);
    assertThat(response.body()).contains("http_requests_total");
    assertThat(response.body()).contains("http_request_duration_seconds_total");
    // byte-for-byte parity with go/python: the space after ';' must survive (Spring's
    // MediaType.parseMediaType would normalize it away — the raw header string preserves it).
    assertThat(response.headers().firstValue("Content-Type").orElse(""))
        .isEqualTo("text/plain; version=0.0.4");
  }

  /**
   * ★ The load-bearing wiring proof: install the file-config live provider, then rewrite the SAME
   * flag file and observe {@code /greeting} flip on the SAME running server with NO restart — the
   * seam flips the REAL endpoint, not a side process.
   */
  @Test
  void greetingLiveFlipOnSameRunningServer(@org.junit.jupiter.api.io.TempDir Path tmp)
      throws Exception {
    Path flagFile = tmp.resolve("flags.json");
    Files.writeString(flagFile, "{\"new_greeting\":false}", StandardCharsets.UTF_8);
    FeatureFlags.setProvider(new FileConfigProvider(flagFile.toString()));

    HttpResponse<String> off = get("/greeting");
    assertThat(off.body())
        .as("pre-flip: the default greeting")
        .isEqualTo("{\"greeting\":\"Hello, world!\"}");

    // Rewrite the SAME file — NO server restart between these two GETs.
    Files.writeString(flagFile, "{\"new_greeting\":true}", StandardCharsets.UTF_8);

    HttpResponse<String> on = get("/greeting");
    assertThat(on.body())
        .as("post-flip: the new greeting, same running server")
        .isEqualTo("{\"greeting\":\"Hello, world! (new)\"}");
  }

  @Test
  void getCarriesSecurityHeadersExactlyOnceAndNeutralServer() throws Exception {
    HttpResponse<String> response = get("/healthz");

    assertThat(response.statusCode()).isEqualTo(200);
    EXPECTED_SECURITY_HEADERS.forEach(
        (name, value) -> {
          List<String> values = response.headers().allValues(name);
          assertThat(values).as("header %s exactly once", name).containsExactly(value);
        });
    assertThat(response.headers().firstValue("Server").orElse(""))
        .as("neutral Server header, no framework/version leak")
        .isEqualTo("reference-app");
  }

  @Test
  void everyNonGetMethodReturns404WithBodyAndSecurityHeaders() throws Exception {
    for (String method : List.of("POST", "PUT", "DELETE", "PATCH", "OPTIONS")) {
      HttpResponse<String> response = send(method, "/greeting", Map.of());

      assertThat(response.statusCode()).as("%s -> 404", method).isEqualTo(404);
      assertThat(response.body()).as("%s body", method).isEqualTo("{\"error\":\"not found\"}");
      EXPECTED_SECURITY_HEADERS.forEach(
          (name, value) ->
              assertThat(response.headers().allValues(name))
                  .as("%s header %s exactly once", method, name)
                  .containsExactly(value));
      assertThat(response.headers().firstValue("Server").orElse(""))
          .as("%s neutral Server header", method)
          .isEqualTo("reference-app");
    }
  }

  @Test
  void headReturns404StatusAndHeadersWithoutBody() throws Exception {
    HttpResponse<String> response = send("HEAD", "/healthz", Map.of());

    assertThat(response.statusCode()).isEqualTo(404);
    assertThat(response.body()).as("HEAD carries no body").isEmpty();
    EXPECTED_SECURITY_HEADERS
        .keySet()
        .forEach(
            name ->
                assertThat(response.headers().firstValue(name))
                    .as("HEAD carries security header %s", name)
                    .isPresent());
    assertThat(response.headers().firstValue("Server").orElse("")).isEqualTo("reference-app");
  }

  @Test
  void unknownGetPathReturnsHardened404() throws Exception {
    HttpResponse<String> response = get("/no-such-route");

    assertThat(response.statusCode()).isEqualTo(404);
    assertThat(response.body()).isEqualTo("{\"error\":\"not found\"}");
    assertThat(response.headers().firstValue("Server").orElse("")).isEqualTo("reference-app");
  }

  @Test
  void validInboundRequestIdIsEchoedIntoSpan() throws Exception {
    String validId = "req-" + UUID.randomUUID().toString();
    HttpResponse<String> response = send("GET", "/healthz", Map.of("X-Request-Id", validId));

    assertThat(response.statusCode()).isEqualTo(200);
    assertThat(spanRequestIdsAfter(response)).contains(validId);
  }

  @Test
  void oversizedInboundRequestIdIsReplacedNotEchoed() throws Exception {
    String oversized = "x".repeat(129); // 129 > the 128-char bound
    HttpResponse<String> response = send("GET", "/healthz", Map.of("X-Request-Id", oversized));

    assertThat(response.statusCode()).isEqualTo(200);
    assertThat(spanRequestIdsAfter(response))
        .as("an oversized inbound id must be replaced by a minted id, never echoed")
        .doesNotContain(oversized);
  }

  /**
   * Reads every emitted span's {@code request_id} attribute from the OTEL trace sink. The sink is
   * configured via the {@code OTEL_TRACE_FILE} env var (surefire, see pom); telemetry is emitted
   * AFTER the response is written, so the request must have completed before this reads.
   */
  private List<String> spanRequestIdsAfter(HttpResponse<String> ignoredCompletedResponse)
      throws IOException {
    String traceFile = System.getenv("OTEL_TRACE_FILE");
    assertThat(traceFile).as("OTEL_TRACE_FILE must be configured (surefire)").isNotNull();
    List<String> lines = Files.readAllLines(Path.of(traceFile), StandardCharsets.UTF_8);
    return lines.stream()
        .map(GreetingIntegrationTest::spanRequestId)
        .filter(id -> id != null)
        .toList();
  }

  private static String spanRequestId(String jsonLine) {
    try {
      JsonNode span = MAPPER.readTree(jsonLine);
      JsonNode requestId = span.path("attributes").path("request_id");
      return requestId.isTextual() ? requestId.asText() : null;
    } catch (IOException e) {
      return null;
    }
  }
}

File: profiles/kotlin/scaffold/src/test/kotlin/com/example/app/GreetingIntegrationTest.kt
package com.example.app

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.server.LocalServerPort
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.file.Files
import java.nio.file.Path
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * Integration: the flag seam + telemetry wiring THROUGH the RUNNING server.
 *
 * Unlike the unit tests (isEnabled / dispatch in isolation), here the flag registry,
 * the [SecurityAndTelemetryFilter], the controllers, and the real embedded Tomcat
 * are exercised together against an ephemeral port (`RANDOM_PORT`) in the SAME
 * process — so JaCoCo measures the spine. Mirrors the go/python integration suites.
 * The live-flip case is the load-bearing proof that the provider seam reaches the
 * REAL endpoint with no restart.
 */
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class GreetingIntegrationTest {
    @LocalServerPort
    private var port: Int = 0

    private val client: HttpClient = HttpClient.newBuilder().version(HttpClient.Version.HTTP_1_1).build()
    private val mapper = jacksonObjectMapper()

    @AfterEach
    fun restoreFloor() {
        // FeatureFlags is a process-wide singleton; never leak an installed provider.
        FeatureFlags.resetProvider()
    }

    private fun base() = "http://127.0.0.1:$port"

    private fun send(
        method: String,
        path: String,
        headers: Map<String, String> = emptyMap(),
    ): HttpResponse<String> {
        val builder = HttpRequest.newBuilder(URI.create(base() + path))
        headers.forEach { (k, v) -> builder.header(k, v) }
        builder.method(method, HttpRequest.BodyPublishers.noBody())
        return client.send(builder.build(), HttpResponse.BodyHandlers.ofString())
    }

    private val expectedSecurityHeaders =
        mapOf(
            "X-Content-Type-Options" to "nosniff",
            "X-Frame-Options" to "DENY",
            "Content-Security-Policy" to "default-src 'none'",
            "Referrer-Policy" to "no-referrer",
        )

    private fun assertSecurityHeadersExactlyOnce(resp: HttpResponse<*>) {
        for ((name, value) in expectedSecurityHeaders) {
            val values = resp.headers().allValues(name)
            assertEquals(listOf(value), values, "header $name must appear exactly once as $value")
        }
        assertEquals("reference-app", resp.headers().firstValue("Server").orElse(null), "Server must be neutralized")
    }

    @Test
    fun `greeting with flag OFF serves the default body`() {
        FeatureFlags.resetProvider()
        val resp = send("GET", "/greeting")
        assertEquals(200, resp.statusCode())
        assertTrue(
            resp.headers().firstValue("Content-Type").orElse("").startsWith("application/json"),
            "content-type = ${resp.headers().firstValue("Content-Type").orElse("")}",
        )
        assertEquals("""{"greeting":"Hello, world!"}""", resp.body())
    }

    @Test
    fun `greeting with flag ON serves the new body`() {
        // The JVM process environment is read-only, so drive ON through the seam.
        FeatureFlags.setProvider { it == "new_greeting" }
        val resp = send("GET", "/greeting")
        assertEquals(200, resp.statusCode())
        assertEquals("""{"greeting":"Hello, world! (new)"}""", resp.body())
    }

    @Test
    fun `healthz returns 200 with status ok`() {
        val resp = send("GET", "/healthz")
        assertEquals(200, resp.statusCode())
        assertEquals("""{"status":"ok"}""", resp.body())
    }

    @Test
    fun `metrics exposes the request counter`() {
        send("GET", "/greeting") // record at least one request
        val resp = send("GET", "/metrics")
        assertEquals(200, resp.statusCode())
        assertTrue(resp.body().contains("http_requests_total"), "metrics missing counter:\n${resp.body()}")
    }

    /**
     * LOCK: the /metrics Content-Type is the cross-profile parity contract, INCLUDING
     * the space after the semicolon (`text/plain; version=0.0.4`). go and python both
     * emit the spaced form; routing through Spring's MediaType round-trip drops the
     * space, so this asserts the RAW value survives through the running server.
     */
    @Test
    fun `metrics Content-Type is the spaced parity contract exactly`() {
        val resp = send("GET", "/metrics")
        assertEquals(200, resp.statusCode())
        assertEquals(
            "text/plain; version=0.0.4",
            resp.headers().firstValue("Content-Type").orElse(null),
            "the /metrics Content-Type must equal the spaced parity contract byte-for-byte",
        )
    }

    /**
     * ★ The load-bearing wiring proof: install the file-config live provider, then
     * rewrite the SAME flag file and observe /greeting flip on the SAME running
     * server with NO restart. Proves the seam flips the REAL endpoint.
     */
    @Test
    fun `greeting live-flips on the same running server with no restart`(
        @TempDir tempDir: Path,
    ) {
        val flagFile = tempDir.resolve("flags.json")
        Files.writeString(flagFile, """{"new_greeting":false}""")
        FeatureFlags.setProvider(fileConfigProvider(flagFile.toString()))

        val off = send("GET", "/greeting")
        assertEquals("""{"greeting":"Hello, world!"}""", off.body(), "pre-flip must serve the default greeting")

        // Rewrite the SAME file — no server restart between these two GETs.
        Files.writeString(flagFile, """{"new_greeting":true}""")
        val on = send("GET", "/greeting")
        assertEquals("""{"greeting":"Hello, world! (new)"}""", on.body(), "post-flip must serve the new greeting")
    }

    @Test
    fun `GET carries the four security headers exactly once and a neutral Server`() {
        val resp = send("GET", "/healthz")
        assertEquals(200, resp.statusCode())
        assertSecurityHeadersExactlyOnce(resp)
    }

    @Test
    fun `every non-GET method returns a hardened 404 with security headers`() {
        for (method in listOf("POST", "PUT", "DELETE", "PATCH", "OPTIONS")) {
            val resp = send(method, "/greeting")
            assertEquals(404, resp.statusCode(), "$method status")
            assertEquals("""{"error":"not found"}""", resp.body(), "$method body")
            assertSecurityHeadersExactlyOnce(resp)
        }
    }

    @Test
    fun `HEAD returns a 404 with security headers and no body`() {
        val resp = send("HEAD", "/healthz")
        assertEquals(404, resp.statusCode())
        assertEquals("", resp.body(), "HEAD must carry no body")
        assertSecurityHeadersExactlyOnce(resp)
    }

    @Test
    fun `unknown GET path returns the hardened JSON 404`() {
        val resp = send("GET", "/nope")
        assertEquals(404, resp.statusCode())
        assertEquals("""{"error":"not found"}""", resp.body())
        assertSecurityHeadersExactlyOnce(resp)
    }

    @Test
    fun `a valid inbound X-Request-Id is echoed into the span`() {
        val validId = "abc-123_valid.ID"
        val got = spanRequestIdFor(mapOf("X-Request-Id" to validId))
        assertEquals(validId, got, "the honoured inbound id must reach the span")
    }

    @Test
    fun `an oversized inbound X-Request-Id is replaced by a minted id`() {
        val bad = "x".repeat(129)
        val minted = spanRequestIdFor(mapOf("X-Request-Id" to bad))
        assertNotEquals(bad, minted, "oversized inbound id must be rejected, not echoed")
        assertEquals(32, minted.length, "minted id must be 32 hex chars")
    }

    /**
     * Drive a GET /healthz with [headers] and return the request_id its emitted span
     * carried. Telemetry is emitted AFTER the response is written (on the worker
     * thread), and spans go to stdout when OTEL_TRACE_FILE is unset — so capture
     * stdout and poll for the span, mirroring the go suite's trace-file poll.
     */
    private fun spanRequestIdFor(headers: Map<String, String>): String {
        val original = System.out
        val buffer = ByteArrayOutputStream()
        System.setOut(PrintStream(buffer, true, "UTF-8"))
        try {
            val resp = send("GET", "/healthz", headers)
            assertEquals(200, resp.statusCode())
            val deadline = System.currentTimeMillis() + 2_000
            while (System.currentTimeMillis() < deadline) {
                lastSpanRequestId(buffer.toString("UTF-8"))?.let { return it }
                Thread.sleep(10)
            }
            fail("no span was emitted to stdout within the timeout")
        } finally {
            System.setOut(original)
        }
    }

    /** Parse ndjson lines; return the request_id of the last valid span, or null. */
    private fun lastSpanRequestId(captured: String): String? {
        var found: String? = null
        for (line in captured.lineSequence()) {
            if (line.isBlank()) continue
            val node: JsonNode =
                try {
                    mapper.readTree(line)
                } catch (_: Exception) {
                    continue
                }
            val attrs = node.get("attributes") ?: continue
            val id = attrs.get("request_id") ?: continue
            if (node.has("trace_id")) found = id.asText()
        }
        return found
    }
}
